assignment2/task2.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: UTF-8 -*-
+
+import math
+
+
+class LCABuilder(object):
+    def __init__(self, graph):
+        self.graph = graph
+        self.root = 0
+        self.two_pow = int(math.ceil(math.log(len(graph), 2)))
+
+        self.time_in = [0] * len(graph)
+        self.time_out = [0] * len(graph)
+        self.parent = [[0] * (self.two_pow + 1) for _ in xrange(len(graph))]
+        self.timer = 1
+
+    def build(self):
+        self._dfs(self.root)
+
+    def _dfs(self, ver, pnt=0):
+        self.time_in[ver] = self.timer
+        self.timer += 1
+
+        self.parent[ver][0] = pnt
+        for index in xrange(1, self.two_pow + 1):
+            prev = self.parent[ver][index - 1]
+            self.parent[ver][index] = self.parent[prev][index - 1]
+        for index in xrange(len(self.graph[ver])):
+            to = self.graph[ver][index]
+            if to == pnt:
+                continue
+            self._dfs(to, ver)
+
+        self.time_out[ver] = self.timer
```

Comment:
Save leaving time from vertex. Need for checking if one vertex in an ancestor of another with O(1) time complexity (this checking is implementing in _check method)